[2주차] 김승완 미션 제출합니다.

Programming-Seungwan opened the pull request:
> # 배포 링크
>
> [배포 링크](https://react-todo-19th-by-react-js-git-progr-7d64b9-seungwans-projects.vercel.app/)

## 📌 구현 기능

1. 지난 주차의 기본 디자인을 적용하여 todo list 기능 구현
2. todo 추가, 완료 전환, 삭제 기능 구현
3. 각 날짜에 맞는 todoList, doneList 불러오기 구현
4. 새로 고침 시 기존의 todo 보존 using localStorage
5. 사용자가 date 타입의 input 달력을 이용하지 않고도 side bar의 직관적인 클릭을 통해서도 날짜 및 관련 UI 변환 가능

## 🧠 느낀 점 및 시간 투자 부분

1. 지난 주의 바닐라 Javascript를 이용할 때보다는 조금 더 체계적인 React의 설계 철학의 이점을 느낀 것 같습니다. 또한 컴포넌트의 개념은 UI의 재사용성을 가능하게 한다는 점이 인상에 깊습니다.
2. 특정 컴포넌트에서 특정 상태가 변화하면 이것이 어디까지 반영되어야 하는지, 그렇다면 상태는 어디에 적용되어야 하는지, 아니면 그냥 일종의 전역 상태 개념으로 만들어버려 prop drilling을 피하고 그때 그때 훅으로 상태를 불러다가 사용하는 방법이 옳은지 고민했습니다.
3. 지난 주에는 막연하게 상태가 없어서 로컬 스토리지를 사용한다고 생각했었는데, 오산이었습니다. 웹 애플리케이션의 기능을 고려했을 때, 사용자가 새로 고침해도 기존에 입력한 todo나 done 아이템들이 유지되어야 하므로 여전히 로컬 스토리지를 사용했어야 한다는 것을 알 수 있었습니다. 이를 아예 DB에 넣어두고 애플리케이션이 로드되고 컴포넌트들이 마운트 될 때마다 사이드 이펙트로 가져오는 방식도 생각해 보았지만, 굉장히 비싼 비동기 연산이므로 차라리 localStorage를 사용하는 편이 낫겠다는 생각이 들었습니다.<br> 이와 같은 방식을 지원하는 라이브러리로는 `Redux-persist`가 있는데 내부적인 로직은 로컬 스토리지를 사용하는 것으로 알고 있습니다.
4. 저는 상태를 최대한 전역적으로 사용하기 위해 contextAPI를 사용했는데, 상태가 필요한 컴포넌트에서는 `useContext()`훅을 사용합니다. 이는 렌더링 시점에 있어서 한번 렌더링이 된 이후에 진행되므로 상태 값을 UI로 렌더링 해주기 위해서는 `&&` 연산자를 적극 활용해야 하면 좋다는 것을 알 수 있었습니다. 이 부분에서 시간을 좀 쓴 것 같습니다.

## ❓ Key Questions

- Virtual-DOM은 무엇이고, 이를 사용함으로
![react-virtualDom-Image](https://github.com/CEOS-Developers/react-todo-19th/assets/125981945/419f8da7-af1e-4001-8de9-8eb3ef2ff7ad)
서 얻는 이점은 무엇인가요?<br>

  브라우저는 CRP(Critical Rendering Path)의 과정을 통해 html과 css 파일을 파싱해 DOM 트리를 만들고, 이를 기반으로 reflow(레이아웃 구성), repaint의 과정을 거쳐 사용자에게 웹 페이지를 보여줍니다. 하지만 위의 과정을 작은 동작 하나 하나가 실행될 때마다 처음부터 다시 시작한다면 시간이 꽤나 오래 걸릴 것이고, 이는 사용자 경험(UX)를 해칠 수 있습니다. <br>
  이를 막기 위해 React는 Virtual DOM이라는 개념을(사실 리액트 뿐만 아니라 vueJS도 이를 사용합니다) 도입합니다. 이는 `알고리즘`적인 방법으로 컴포넌트 트리를 기존의 것(snapshot)과 비교하여 바뀐 부분만 dom에서 갈아끼워준다는 개념입니다(갈아 끼워주는 것은 프로그래머가 아닌 React가 대신 해줍니다). 하지만 모든 컴포넌트 노드들을 일일히 비교하면 대강 **O(n^3)**의 시간 복잡도를 보이기에 여전히 긴 시간으로 UX를 해칠 수 있습니다.<br>
  이에 React는 휴리스틱 알고리즘을 적용하는데, 그 원리는 다음과 같습니다.

  1. 아예 `element type`이 다른 경우에는 해당 노드로부터의 subtree를 모두 폐기하고 다시 생성합니다.
  2. `element type`은 같은데 노드의 속성(이벤트 핸들러, css 속성 등)이 다른 경우에는 동일한 속성은 유지하고 변경된 속성만 갈아끼워 줍니다.
  3. `자식 요소`가 바뀌는 경우는 재귀적으로 1 단계로 돌아가서 알고리즘을 실행합니다.
  4. 추가적 효율성을 위해 기존에 존재하는 노드가 변경되지 않았을 경우의 불필요한 리렌더링을 막으려면 요소에 `key` 속성을 부여하면 됩니다 => React 컴포넌트의 리스트 렌더링 시 고유한 `key` 값을 부여하는 이유

  > 이와 같은 알고리즘을 Virtual Dom의 `diffing algorithm`이라고 합니다. 이는 렌더링 시간을 단축하여 사용자의 경험을 개선할 수 있습니다 => 이를 통틀어 재조정(**Reconciliation**)이라고 함

- 미션을 진행하면서 느낀, React를 사용함으로서 얻을수 있는 장점은 무엇이었나요?<br>
  동일한 과제를 vanilla JS가 아닌, React를 사용하면서 느낀 장점은 다음과 같습니다.

  1. 중복되는 UI 요소를 매번 선언해주는 것이 아니라, 함수형 컴포넌트로 만들어주어 필요할 때마다 import하여 사용할 수 있다.
  2. **"상태"**를 이용하여 컴포넌트 간에 공유가 가능하게 하며 그 중 한 컴포넌트에서 상태를 변경하면 관련된 컴포넌트들이 이를 구독합니다.
  3. `useState`, `useEffect`와 같은 hook들을 사용하여 리렌더링 후에도 기존의 상태를 유지하거나 변경된 상태가 UI에 적용되었고, 한번 렌더링이 된 이후에 실행하고 싶은 비동기 로직이나 작업들은 `side Effect`로서 처리합니다.
  4. 브라우저 익스텐션에 React 관련 확장을 설치해 특정 요소가 어느 컴포넌트에서 비롯된 것인지, 그리고 다양한 사항들을 `profiler`를 통해 살펴볼 수 있었습니다.
     <br>

     그 이외에도 React를 사용하면 좋은 점들이 매우 많다는 것을 알고 있습니다.

  5. 기존의 **SSR(Server Side Rendering)** 방식을 따르면 초기 로딩 속도는 다소 빠를 수 있지만 페이지 전환 시, 뚝뚝 끊기는 현상을 변경되는 컴포넌트만 리렌더링하기에 개선할 수 있습니다.
  6. 다양한 로직들을 따로 커스텀 훅으로 분리하는 등의 관심사를 나눌 수 있습니다.

- React에서 상태란 무엇이고 어떻게 관리할 수 있을까요?<br>
![react-class-component-life-cycle](https://github.com/CEOS-Developers/react-todo-19th/assets/125981945/75172d8f-66e5-4681-8b18-4a31ab6956d9)

  기존의 바닐라 JS에서는 변수나 리터럴 값들을 UI 렌더링에 사용하곤 했습니다. 하지만 React 라이브러리는 이전의 class 방식과 functional 방식 모두 렌더링이라는 방법을 사용합니다. 전자에서는 `render()`라는 메서드로, 후자는 컴포넌트 함수의 `return` 값으로 이를 구현합니다. 하지만 이때 변수를 사용하면 메서드나 함수가 실행될 때 변수는 또 다시 초기화될 뿐입니다. <br>
  따라서 React는 "상태"라는 개념을 도입합니다. 이는 리-렌더링 시에도 유지되는 것으로서 React 라이브러리의 출발점이라고 할 수 있습니다. `const [count, setCount] = useState(0)`와 같은 방식으로 선언할 수 있으며 객체 구조 분해의 첫 번쨰 요소가 바로 상태이고 두 번째 요소는 이를 변화시킬 수 있는 함수입니다. 기존의 변수처럼 재할당과 같은 방식은 상태에는 적용되지 않으면 오로지 상태 변경 함수(여기에서는 setCount)를 통해 이루어져야만 합니다. 상태를 선언할 수 있는 방식은 다음과 같습니다. <br><br>

  1. `useState()` 훅 : 윗 문단에서 설명
  2. `useReducer()` 훅 : 기존의 `useState()`가 너무 복잡한 로직을 가질 때 활용할 수 있습니다. 이는 Javascript 배열의 `reduce()` 메서드의 원리를 생각하면 좋습니다.
     > 기존의 것을 받아 새로 업데이트 한다.

  ```
  const reducer = (state, action) => { //state는 기존의 상태이다
  	console.log("Reducer activate!");

  	if (action.type === "plus") state++;
  	else if (action.type === "minus") state--;
  }
  const [money, dispatch] = useReducer(reducer, 0);
  dispatch({type : "plus"});

  ```

  3. `useContext()` 훅: createContext() 함수로 컨텍스트를 만든 뒤, 사용할 상태를 useState() 훅으로 설정해 만든 컨텍스트의 provider 컴포넌트의 value 값으로 전달한다. 후에 사용하는 컴포넌트에서는 `useContext(실행하는 컨텍스트명)`의 방식으로 사용하면 된다.
     <br>
     하지만 위의 방식들을 이용하면 contextAPI를 제외한 나머지 두 방식은 prop drilling으로 일일히 속성을 전달하고, 이어 받아야 하므로 다소 번거로울 수가 있다. 뿐만 아니라 React는 부모 컴포넌트가 리렌더링 되면 그 사이에 있는 불필요한 자식 컴포넌트까지 함꼐 리렌더링 되므로 퍼포먼스의 저하로 UX를 해칠 가능성도 있다. 따라서 다양한 라이브러리를 활용하여 이를 관리할 수 있는데, 방식은 다음과 같다.

  - Redux : `flux` 패턴에 기반하면 이는 단방향으로 데이터가 흐름을 의미한다. `useReducer()` 훅과 작동 원리는 매우 유사한데, 중앙의 저장소(store)에 상태들이 모여있다. 코드적으로 본다면 `action` 객체를 `dispatcher` 함수의 인자로 전달하여 트리거하면 등록해준 리듀서 함수의 로직에 따라 상태가 변경되고 이를 구독하는 컴포넌트들만 리렌더링 되는 방식이다. 하지만 Redux 라이브러리의 경우 간단한 상태 관리를 하는 경우를 고려하면 보일러 플레이트가 너무 크고 비동기적인 상태 관리를 `Thunk`나 `saga` 모듈을 통해 할 수 있지만 다소 복잡하다고 할 수 있다.
  - Recoil : meta에서 만든 atomic 디자인 패턴에 입각하여 만든 상태관리 라이브러리로서, 본질적 상태인 `atom`과 이로부터 파생된 `selector` 개념을 통해 상태를 관리할 수 있다. Redux보다 확연하게 보일러 플레이트가 적으며, 이를 더 줄인 것이 Jotai 라이브러리라고 생각하면 된다.
  - React(Tanstack) Query & SWR : 위의 상태 관리 라이브러리들은 비동기적인 호출로 인한 상태의 반영에 있어서 다소 어려움을 가지고 있다. 두 라이브러리는 이를 `useQuery()`, `useMutation()` 과 같은 훅을 이용해 관리할 수 있다.

- Styled-Components 사용 후기 (CSS와 비교)<br>
  기존의 Scss나 Sass와 같은 css 전처리기의 방식도 있지만 Styled-components 라이브러리는 css-in-js 방식으로 분류되는 방법입니다. 이는 React와 같은 컴포넌트 방식의 프론트엔드 기술과 잘 어울리는데, Javascript 코드(jsx도 js로 보겠습니다) 내부에 코드를 css 코드를 colocate 할 수 있다는 장점이 있습니다.<br>
  css 코드를 javascript 파일 외부에 위치시키면 매번 스타일링을 하기 위해 프로젝트 디렉터리를 찾아다녀야 하는 번거로움이 있고, `module.css` 방식이 아닌 plain css 방식으로 진행하면 클래스명의 충돌 휴먼 에러가 발생할 수 있습니다. <br>
  하지만 Styled-Components 라이브러리는 클래스명 충돌과 같은 문제를 자체적으로 해결해주기 때문에 걱정할 필요가 없어서 좋았습니다.

Programming-Seungwan committed:
Feat : first project settings commit

1. styled-components 라이브러리 추가 설치
2. 과제 기능 구현 문서 작성
3. 컴포넌트를 분리할 디렉터리와 유용한 함수들을 따로 빼놓을 디렉터리 분리

Programming-Seungwan committed:
Chore : reset.css 파일 설정

1. 기본적인 vendor 브라우저 적용 css 무효화
2. scroll-bar 관련 설정

Programming-Seungwan committed:
Chore : make default total layout

Programming-Seungwan committed:
Chore : make default main content area layout

Programming-Seungwan committed:
Feat : body 기본 레이아웃 설계

Programming-Seungwan committed:
Style : apply airbnb code convention in eslint and prettier config

Programming-Seungwan committed:
Chore : make body content area by flex box

Programming-Seungwan committed:
Chore : css in body upper(date picker and todo text input area)

Programming-Seungwan committed:
Build : installed font-awesome react component

1. app container 자체를 flex box로 만들기보다는 통상적인 maring: 0 auto 속성으로 맞춰줌
2. font awesome의 아이콘을 사용하기 위해서 관련 라이브러리 설치

Programming-Seungwan committed:
Chore : make default body upper layout

Programming-Seungwan committed:
Feat : contextAPI를 이용해서 사용자가 날짜를 바꾸면 전체 상태에 적용 가능

Programming-Seungwan committed:
Feat : 할일 추가 기능 완성

Programming-Seungwan committed:
Feat : 사용자가 등록한 할일을 로컬 스토리지에서 받아와 상태와 동기화하여 UI에 렌더링 해주기

1. 상태를 useContext() 훅을 이용해 진행하기 때문에 첫 마운트 이후에 제대로 진행된다. 따라서 && 연산자를 사용한다.

Programming-Seungwan committed:
Fix : todo Item에 key값을 부여하여 렌더링 최적화

Programming-Seungwan committed:
Feat : 체크 박스를 누르면 넘어가는 todo에 있던 것이 done으로 넘어가는 로직을 구현

Programming-Seungwan committed:
Feat : done UI first make

Programming-Seungwan committed:
Feat : todo list에서 삭제 버튼을 누르면 삭제되도록 만듦

Programming-Seungwan committed:
Feat : make done list functionality

Programming-Seungwan committed:
Feat : 날짜를 바꾸면 관련된 상태에 적용하여 UI 변경까지 렌더링

Programming-Seungwan committed:
Chore : todoContentArea와 doneContentArea 높이 및 overflow 스크롤 설정

Programming-Seungwan committed:
Feat : make footer copyright text

Programming-Seungwan committed:
Feat : make side bar on pc environment

1. 날짜를 클릭하면 해당 일의 일정으로 바뀜

Programming-Seungwan committed:
Feat : task done UI make

1. 일부 css도 추가

Programming-Seungwan committed:
Fix : make additional styling

1. 모바일 디바이스에서는 todo와 done 박스가 가로가 아니라 세로로 보이도록 만들어줌

Programming-Seungwan committed:
Refactor : 불필요한 주석 제거

Programming-Seungwan committed:
Build : delete unnecessary devDependencies modules

Dahn12 commented on src/components/AppContainer.jsx:
```diff
@@ -0,0 +1,66 @@
+        : JSON.parse(localStorage.getItem(`${selectedDayString}done`));
+    setSelectedDayTodoList(todo);
+    setSelectedDayDoneList(done);
+  }, []);
```
저두 useState 보다 useEffect로 구현하고 싶었는데 이렇게 해주면 되겠네요! 좋습니다!

ddhelop replied:
useEffect 훅이 실행될 때 의존성 배열을 빈 배열로 설정하여 컴포넌트가 마운트될 때만 실행되도록 하면 selectedDayString 상태가 변경될 때 localStorage에서 해당 날짜의 할 일과 완료된 일을 다시 불러와야 해서 의존성 배열에 selectedDayString을 포함시키면 electedDayString이 변경될 때마다 useEffect가 실행되어 해당 날짜의 할 일과 완료된 일을 localStorage에서 불러와서 불필요한 동작이 감소할 것 같아요!

```suggestion
useEffect(() => {
  const todo = JSON.parse(localStorage.getItem(`${selectedDayString}todo`)) || [];
  const done = JSON.parse(localStorage.getItem(`${selectedDayString}done`)) || [];
  setSelectedDayTodoList(todo);
  setSelectedDayDoneList(done);
}, [selectedDayString]);
```

Dahn12 commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/MainContentBodyLower.jsx:
```diff
@@ -0,0 +1,17 @@
+import styled from 'styled-components';
+import OneDayContainer from './OneDayConatiner/OneDayContainer';
+
+const StyledMainContentBodyLower = styled.div`
+  height: 70%;
+  width: 100%;
+  display: flex;
+  justify-content: space-evenly;
+`;
```
각 영역을 flex box를 통해 배치하셨는데, 승완님 todo를 보면 grid로 구현하는 것도 좋을 거 같다는 생각이 들었습니다. 제가 많이 도움 받았던 [블로그글](https://studiomeal.com/archives/533)입니다!

Dahn12 commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/DoneContainer/DoneContentArea.jsx:
```diff
@@ -0,0 +1,26 @@
+`;
+
+export default function DoneContentArea() {
+  const { selectedDayDoneList } = useContext(ThemeContext);
```
저도 이번 미션을 하면서 props 전달 때문에 애를 많이 먹었는데 useContext에 대해 알아갑니다!

ddhelop commented on src/components/AppContainer.jsx:
```diff
@@ -0,0 +1,66 @@
+    min-width: 100%;
+    padding: 0 5%;
+    background-color: #bcdbad;
+    height: 100dvh;
```
처음보는 단위라서 신기하네요! 이 단위를 사용한 이유 궁금해요

Programming-Seungwan replied:
아 높이 설정을 할 떄 그냥 `vh` 단위를 사용하면 아이폰의 사파리나 크롬 브라우저를 통해 볼 때, 주소창이 전체를 가리는 경우가 발생하는 것 같더라구요! 이를 dvh(Dynamic Viewport Height) 속성으로 막아줄 수 있다고 해서 사용해봤습니다! [관련 링크](https://abcdqbbq.tistory.com/104)를 첨부할게요 :)

ddhelop commented on src/components/AppContainer.jsx:
```diff
@@ -0,0 +1,66 @@
+  }, []);
+
+  return (
+    <ThemeContext.Provider
+      value={{
+        selectedDayString,
+        setSelectedDayString,
+        selectedDayTodoList,
+        setSelectedDayTodoList,
+        selectedDayDoneList,
+        setSelectedDayDoneList,
+      }}
+    >
+      <StyledAppContainer>
+        <SideBar />
+        <AppMainContent />
+      </StyledAppContainer>
+    </ThemeContext.Provider>
```
컴포넌트로 각각 나눠서 구조를 잡은거 나중에 관리하기 정말 편할 것 같아요👍 
그치만 개인 취향이긴 하지만 프로젝트 사이즈에 비해 컴포넌트를 너무 쪼개서 조금 더 큰 단위로 쪼개는 것도 좋은 방법인 것 같아요

Dahn12 commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/TodoContainer/TodoItemUI.jsx:
```diff
@@ -0,0 +1,102 @@
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  height: 20%;
+  min-height: 20%;
```
<img width="404" alt="스크린샷 2024-03-24 오후 4 56 46" src="https://github.com/CEOS-Developers/react-todo-19th/assets/119108655/507624ca-6d35-4f25-909a-e6f1c2cc41b2"> flex 내부에서 높이를 주게되면 이렇게 글자가 넘어가도 높이는 유지되는 오류가 있습니다! 저도 이 부분으로 고민을 많이 했었는데, 저는 특정 값을 주지 않고 위아래 margin 값을 주어서 택스트가 들어가는 div 기준으로 위아래 간격을 주는 방식으로 해결했습니다! 하지만 더 좋은 방법이 있을 거 같더라구요..스터디 시간에 이야기해봐요!

leejin-rho replied:
저도 이부분 발견했는데 height: fit-content를 주는 등 높이가 동적으로 변하게 하는 건 어떨까요??

Dahn12 commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/DoneContainer/DoneItemUI.jsx:
```diff
@@ -0,0 +1,109 @@
+  const { selectedDayString, setSelectedDayTodoList, setSelectedDayDoneList } =
+    useContext(ThemeContext);
+
+  const localStorageTodoKey = `${selectedDayString}todo`;
+  const localStorageDoneKey = `${selectedDayString}done`;
```
각 날짜별로 불러올 수 있도록 구현하신거 정말 대단하신거 같아요....👍

ddhelop commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/DoneContainer/DoneItemUI.jsx:
handleClickRollBackArrowIcon과 handleClickTrashCanIcon 함수에서 localStorage의 데이터를 처리하는 로직이 중복되어 있어서 나중에 오류가 생기면 처리하기 어려울 것 같아서 두 함수의 공통 로직을 합치도록 코드를 구성해도 좋을 것 같아요

Dahn12 commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/DoneContainer/DoneItemUI.jsx:
```diff
@@ -0,0 +1,109 @@
+      localStorage.getItem(localStorageDoneKey)
+    );
+
+    const prevLocalStorageDoneIndex = prevLocalStorageDoneData.indexOf(text);
+    prevLocalStorageDoneData.splice(prevLocalStorageDoneIndex, 1);
```
이 과정을 줄이기 위해서 객체 형태로 저장하는 방법도 고려해보시면 좋을거 같아요! 저도 지난 미션 발표를 참고해서 객체형태로 저장했더니 훨씬 직관적이고, 배열 형태에서 다루기도 더 편했던거 같아서 추천드립니다!

Dahn12 commented on src/components/SideBar-component/SideBar.jsx:
```diff
@@ -0,0 +1,70 @@
+export default function SideBar() {
+  const { selectedDayString } = useContext(ThemeContext);
+
+  const dateArray = [];
+  for (let i = -2; i <= 2; i++) {
+    dateArray.push(getDateString(selectedDayString, i));
+  }
```
한 화면에 5개의 날짜가 보이도록 구현하셨는데, overflow에서 scroll을 적용하면 모든 날짜를 보여줄 수 있을 거 같다는 생각이 들었습니다. 하지만 이부분은 개인 취향이니까요...!

ddhelop commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/TodoContainer/TodoItemUI.jsx:
```diff
@@ -0,0 +1,102 @@
+
+  function handleClickCheckIcon() {
+    const prevLocalStorageData = JSON.parse(
+      localStorage.getItem(localStorageTodoKey)
+    );
+    const prevItemLocalStorageIndex = prevLocalStorageData.indexOf(text);
+    prevLocalStorageData.splice(prevItemLocalStorageIndex, 1);
+    localStorage.setItem(
+      localStorageTodoKey,
+      JSON.stringify(prevLocalStorageData)
+    );
+
+    const prevLocalStorageDoneData =
+      localStorage.getItem(localStorageDoneKey) === null
+        ? []
+        : JSON.parse(localStorage.getItem(localStorageDoneKey));
+    prevLocalStorageDoneData.push(text);
+    localStorage.setItem(
+      localStorageDoneKey,
+      JSON.stringify(prevLocalStorageDoneData)
+    );
+
+    setSelectedDayTodoList(prevLocalStorageData);
+    setSelectedDayDoneList(prevLocalStorageDoneData);
+  }
+
+  // 로컬 스토리지에서 삭제하고 UI에서도 삭제를 해야함
+  function handleClickTrashIcon() {
+    const prevLocalStorageData = JSON.parse(
+      localStorage.getItem(localStorageTodoKey)
+    );
+
+    const prevLocalStorageIndex = prevLocalStorageData.indexOf(text);
+    prevLocalStorageData.splice(prevLocalStorageIndex, 1);
+    localStorage.setItem(
+      localStorageTodoKey,
+      JSON.stringify(prevLocalStorageData)
+    );
+    setSelectedDayTodoList(prevLocalStorageData);
+  }
```
정확하진 않지만 이런느낌으로 리펙토링하면 코드 수를 줄일 수 있고, 로컬스토리지 접근도 간소화 할 수 있는 아이디어 일 것 같아요

```suggestion
// 로컬 스토리지 업데이트 로직을 별도의 함수로 분리
function updateLocalStorageItem(key, updateFunction) {
  const data = JSON.parse(localStorage.getItem(key)) || [];
  updateFunction(data);
  localStorage.setItem(key, JSON.stringify(data));
}

// 개선된 handleClickCheckIcon 함수
function handleClickCheckIcon() {
  updateLocalStorageItem(localStorageTodoKey, (data) => {
    const index = data.indexOf(text);
    if (index > -1) data.splice(index, 1);
  });
  updateLocalStorageItem(localStorageDoneKey, (data) => {
    data.push(text);
  });

  // 상태 업데이트 로직
}

// 개선된 handleClickTrashIcon 함수
function handleClickTrashIcon() {
  updateLocalStorageItem(localStorageTodoKey, (data) => {
    const index = data.indexOf(text);
    if (index > -1) data.splice(index, 1);
  });
```

Dahn12 reviewed (commented):
승완님 이번 미션도 정말 수고 많았어요! 날짜 별로 local storage를 구현하셨다는 점이 정말 인상깊었습니다. 또한 컴포넌트도 굉장히 세분화해서 나누어 주셨던데, 저도 여전히 컴포넌트를 나누는 기준에 대해 고민하고있는 중이라 스터디 시간에 더 이야기해보면 좋을거 같아요..! 다음 미션도 같이 화이팅 해봐요!👍

ddhelop commented on src/components/MainContent-component/MainContent-Body/MainContentBodyUpper/DatePickerInput.jsx:
```diff
@@ -0,0 +1,59 @@
+  font-weight: 500;
+`;
+
+export default function DatePickerInput(second) {
```
second 파라미터를 받고 있는데, 사용하지 않는 것 같아요!

ddhelop commented on src/components/MainContent-component/MainContent-Body/MainContentBodyUpper/TodoInput.jsx:
```diff
@@ -0,0 +1,110 @@
+  const doneNum = selectedDayDoneList ? selectedDayDoneList.length : 0;
+
+  return (
+    <StyledInputFormContainer onSubmit={handleSubmitTodoInputForm}>
+      <StyledInputForm>
+        <StyledTodoTextInput
+          type="text"
+          placeholder="Add your new Todo List!"
+          ref={todoInputRef}
+        />
+        <StyledTodoSubmitButton
+          icon={faCirclePlus}
+          onClick={handleClickPlusLogoButton}
+        />
```
handleSubmitTodoInputForm 함수와 handleClickPlusLogoButton 함수가 모두 존재하며, 폼 제출 시 handleSubmitTodoInputForm이 handleClickPlusLogoButton을 호출해서 불필요한 중복 호출을 합니다.
폼 제출 이벤트(onSubmit)만 사용하면 중복 호출이 안 될 것 같아요.

leejin-rho commented on src/components/SideBar-component/DateCard.jsx:
```diff
@@ -0,0 +1,51 @@
+  }
+  return (
+    <StyledDateCard onClick={handleClickDateCard}>
+      <FontAwesomeIcon icon={faCalendar} />
```
날짜카드를 만들어 구현해주신 점 정말 흥미롭게 만들어주셨네요~!

leejin-rho commented on src/components/MainContent-component/MainContent-Body/MainContentBodyLower/OneDayConatiner/TodoContainer/TodoContainer.jsx:
```diff
@@ -0,0 +1,30 @@
+  display: flex;
+  flex-direction: column;
+  flex-shrink: 0;
+  width: 49%;
```
모두 반응형으로 신경써서 구현해주신 점 좋습니다~!

leejin-rho commented on src/components/MainContent-component/MainContent-Footer/MainContentFooter.jsx:
```diff
@@ -0,0 +1,20 @@
+export default function MainContentFooter() {
+  return (
+    <StyledMainContentFooter>
+      <p>All rights reserved © Seungwan, Kim 2024</p>
```
요런 디테일도 재밌네용 ㅎㅎㅎ

leejin-rho reviewed (commented):
안녕하세요 프론트 운영진 노이진입니다!! 이번과제도 수고하셨습니다~!! 엠티는 잘 다녀오셨나용?!ㅎㅎㅎ 다음과제는 합동과제인 만큼 더 재미있을 거에요..! 그리고 디자인 신경도 안써도 됩니다 ㅎㅎㅎ 비대면 스터디에서 만나요~

ddhelop reviewed (commented):
엄청난 정성과 시간을 쏟으신게 코드에서 보여요.. 엄청난 열정.
컴포넌트 잘게잘게 쪼개신것도 인상적이였어요. 고생하셨어요

개인 취향이긴 하겠으나 함수나 변수의 이름이 길어서 좀 더 직관적이게 네이밍을 하면 함수나 변수가 무엇을 하는지 명확히 알 수 있을 것 같아요.

Programming-Seungwan committed:
Refactor : change context name by contextAPI

1. 기존의 contextAPI를 통해서 만든 상태명을 ThemeContext라고 너무 말도 안되게 지었는데 이를 TodoContext라는 좀 더 semantic한 이름으로 변경함

Programming-Seungwan committed:
Fix : prevetn todo or done UI overflow from container

Programming-Seungwan committed:
Fix : selectedDayString 상태가 바뀔 때마다 다시 로컬스토리지로부터 상태를 불러와서 이와 연관된 상태들을 연쇄적으로 바꿔주는 로직을 AppContainer의 useEffect() 훅을 통해 구현함

Programming-Seungwan committed:
Fix : form 태그를 사용하여 제출을 하게 만들면 button 요소에 onClick() 핸들러 함수를 달아주지 않아도 onSubmit 속성으로 달린 핸들러 함수가 트리거 되기에 불필요하게 사용할 필요가 없다.
